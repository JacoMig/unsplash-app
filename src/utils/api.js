import Unsplash, { toJson } from 'unsplash-js';

const unsplash = new Unsplash({
    accessKey: process.env.REACT_APP_ACCESS_KEY,
    secret: process.env.REACT_APP_SECRET_KEY,
    headers: new Headers({
        'Accept': 'application/json',
        'Content-Type': 'application/json',
        'Cache-Control': 'no-cache',
    }),
    timeout: 500
});

export const getPhotos = (options) => {
    //console.log(process.env.REACT_APP_ACCESS_KEY)
    const request = unsplash.search.photos(options.query, options.page_start, 12 ,{ orientation: "landscape" }).then(toJson)
    return request
    
}