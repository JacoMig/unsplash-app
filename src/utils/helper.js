import React from 'react';
import { Router, Route} from 'react-router-dom'
import { createMemoryHistory } from 'history'
// import { render, cleanup, fireEvent } from '@testing-library/react';

// Helper function
export function componentWithRouterMatch(
    Ui,
    {
      path = "/",
      route = "/",
      history = createMemoryHistory({ initialEntries: [route] })
    } = {}
  ) {
    return (
      
        <Router history={history}>
          <Route 
            path={path} 
            render={routeProps =>
              <Ui history={history} match={routeProps.match} />
            }
          />
        </Router>
      
    );
  }