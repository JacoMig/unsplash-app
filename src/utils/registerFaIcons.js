import { library } from '@fortawesome/fontawesome-svg-core'
import { faSearch, faUser, faTags, faThumbsUp, faCalendarAlt } from '@fortawesome/free-solid-svg-icons'


export default function registerIcons() {
	library.add(faSearch, faUser, faTags, faThumbsUp, faCalendarAlt); 
}