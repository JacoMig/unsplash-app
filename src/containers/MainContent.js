import React, { useEffect, useState} from 'react';
import {getPhotos} from '../utils/api';
import {Row, Col, Container, Spinner} from 'reactstrap';
import PhotoGallery from '../components/PhotoGallery';

let count = 0

const MainContent = (props) => {
    
    const {match} = props;
    // const query = match.params.query ? match.params.query : ""
    const [query, setQuery] = useState("")
    const [photos, setPhotos] = useState([]);
    const [totalPages, setTotalPages] = useState(0);
    const [endLoading, setEndLoading] = useState(false);
    const [loading, setIsLoading] = useState(false);
    const loader = React.useRef();
    //const queryRef = React.useRef(query);
    
    const loadPhotos = async () => {
        count += 1
       // setIsLoading(true)
        if(count === totalPages) {
            setEndLoading(true);
            return
        }else {
            getPhotos({query: match.params.query, page_start: count})
                .then(res => {
                    setPhotos(state => [...state, ...res.results]);
                    count === 1 && setTotalPages(res.total_pages);
                
                });
           // setIsLoading(false)
        }
    }

   

    useEffect(() => {
        if(match.params.query !== query){
            setPhotos([]);
            setQuery(match.params.query)
            console.log('clean photos Array')
            //setIsLoading
            // fetchData()  
        }else {
            
        }
    }, [match.params.query])

    useEffect(() => {
        if(query !== ""){
            const handleObserver = async (entities) => {
                const target = entities[0];
                console.log(entities)
              
                if (target.isIntersecting) {   
                    console.log('loadPhotos Observer Lazy');
                    loadPhotos();
                }
            }
            var options = {
                root: null,
                rootMargin: "5px",
                threshold: 1.0
            };
            const observer = new IntersectionObserver(handleObserver, options);
            if (loader.current) {
                observer.observe(loader.current);
                // console.log(loader.current)
            }
            //console.log(queryRef)
        }
        
    }, [query])
    

   /*  useEffect(() => {
        if(query !== ""){
            const fetchData = async () => {
                setIsLoading(true);
                setPhotos([]);
                count=0
                await loadPhotos();
                setIsLoading(false)
                console.log('initial fetchdata')
            }
            fetchData()    
        }
    },[query])  */

    

    return (
        <Container fluid>
             <Row>
                {!loading &&  photos.length > 0 && 
                    <PhotoGallery endLoading={endLoading} loadPhotos={loadPhotos} photos={photos} />
                }
                {!loading &&  photos.length === 0 && 
                    <Col col={12}><h6>Nessun risultato trovato per: <i>{query}</i></h6></Col>
                }
                {/*  photos.length > 0 && */
                    <div className="m-auto"  ref={loader}>
                        {loading && <Spinner animation="border" variant="info" />}
                    </div>
                }
            </Row>
        </Container>
    );
};

export default MainContent;