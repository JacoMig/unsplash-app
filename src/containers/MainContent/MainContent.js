import React, { useEffect, useState} from 'react';
import {getPhotos} from '../../utils/api';
import {Row, Col, Container, Spinner} from 'reactstrap';
import PhotoGallery from '../../components/PhotoGallery/PhotoGallery';

let count = 0

export const loadPhotos = (query, setPhotos, setEndLoading) => {
    count += 1
    return getPhotos({query, page_start: count})
        .then(res => {
            setPhotos(state => [...state, ...res.results]);
            if(count === res.total_pages) {
                setEndLoading(true);
            }
    }).catch(e => console.log(e))
}

const MainContent = (props) => {
    
    const {match} = props;
    const query = match.params.query ? match.params.query : ""
    const [photos, setPhotos] = useState([]);
    const [endLoading, setEndLoading] = useState(false);
    const [loading, setIsLoading] = useState(false);
 
   
   /*  const loadPhotos = () => {
       
    } */

   
    useEffect(() => {
        if(query !== ""){
            const fetchData = async () => {
                setIsLoading(true);
                setPhotos([]);
                count = 0;
                await loadPhotos(query, setPhotos, setEndLoading);
                setIsLoading(false)
                setEndLoading(false);
                console.log('initial fetchdata')
            }
            fetchData()    
        }
    },[query])

    

    return (
        <Container>
             <Row>
                {/* photos.length > 0 && <p className="testP" id="testP">{photos[0].id}</p> */}
                {!loading &&  photos.length > 0 && 
                    <PhotoGallery endLoading={endLoading} loadPhotos={() => loadPhotos(query, setPhotos, setEndLoading)} photos={photos} />
                }
                {!loading &&  photos.length === 0 && 
                    <Col col={12}><h6>Nessun risultato trovato per: <i>{query}</i></h6></Col>
                }
                {loading && <Spinner animation="border" variant="info" />}
            </Row>
        </Container>
    );
};

export default MainContent;