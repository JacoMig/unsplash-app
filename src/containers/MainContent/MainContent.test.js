import * as React from 'react';
import ReactDOM from 'react-dom';
import { render, cleanup, fireEvent, waitForElement } from '@testing-library/react';
import {componentWithRouterMatch} from '../../utils/helper';
import { createMemoryHistory } from 'history'
import MainContent from './MainContent'
import registerFaIcons from '../../utils/registerFaIcons';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import * as fetchApi from '../../utils/api';

import {loadPhotos} from './MainContent';
Enzyme.configure({ adapter: new Adapter() })

registerFaIcons()

let wrapper;
// let history;
const myquery = "cat"
const componentProps = {
    route: "/",
    path: "/:query?",
    match: {params: {query: myquery}}
}

describe("MainContent testing", () => {

  /* const setPhotos = jest.fn();
  const useStateSpy = jest.spyOn(React, "useState");
  useStateSpy.mockImplementation(photos => [photos, setPhotos]);  */
  
  beforeEach(() => {
    wrapper = Enzyme.mount(Enzyme.shallow(<MainContent {...componentProps} />).get(0))
  }) 

  afterEach(() => {
    jest.clearAllMocks();
    cleanup;
  });


  test("Main Content matches Snapshot", () => {
     expect(wrapper).toMatchSnapshot()
  }) 

  test("it renders Main Content without crash", () => {
    const div = document.createElement("div");
    ReactDOM.render(componentWithRouterMatch(MainContent, {
      ...componentProps
    }), div);
  }) 
  
  
  jest.mock('../../utils/api');
  
  test("Check if Fetching Api returns something", async () => {
    await fetchApi.getPhotos({query: 'woman', page_start: 1}).then(data => {
      expect(data.results.length > 0).toBe(true);
    })  
  }) 
  
  
  const setPhotos = jest.fn()
  test("Check if setPhotos is called after fetch Api", async () => {
    await loadPhotos('cat', setPhotos)
    expect(setPhotos).toHaveBeenCalled()
   
  })

})
