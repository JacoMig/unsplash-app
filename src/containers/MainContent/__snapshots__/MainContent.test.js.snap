// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainContent testing Main Content matches Snapshot 1`] = `ReactWrapper {}`;
